notes/smart-pointers-1.md

```diff
@@ -0,0 +1,244 @@
+
+Let's take a look at 3 kinds of smart pointers
+
+---
+## `std::unique_ptr`
+
+A `unique_ptr` is a kind of smart pointer
+- Use this for managing resources with exclusive ownership semantics
+- Very little overhead, can be used even when memory is limited
+
+Since it is a kind of smart pointer, a unique pointer will take ownership of a pointer to an object
+
+The `unique_ptr`'s destructor will invoke `delete` on the owned pointer when the `unique_ptr` object is destroyed
+
+Example:
+```C++
+#include <iostream> // std::cout, std::endl
+#include <memory> // std::unique_ptr
+
+void foo(){
+  std::unique_ptr<int> x(new int(333));
+  // You can dereference x as if it were a normal pointer
+  *x += 20;
+  std::cout << *x << std::endl;
+
+  // No leaks, even though we never used delete!
+  // When x falls out of scope, we run delete on its owned ptr
+}
+
+```
+
+---
+## `unique_ptr` Operations
+
+For unique pointer `x`:
+
+`x.get()` returns a pointer to the pointed-to object
+
+`*x` returns value of pointed-to object
+- Basically normal dereference
+
+`x.reset(ptr)`: run `delete` on the current owned pointer, then store a new one, `ptr`
+
+`x.release()` returns the pointer, sets wrapped pointer to `nullptr`
+- Releases responsibility for explicitly freeing back to user
+
+Example:
+```C++
+#include <memory> // unique_ptr
+#include <iostream> // cout
+#include <cstdlib> // EXIT_SUCCESS
+
+int main(){
+  // x is a unique pointer that owns an int*
+  std::unique_ptr<int> x(new int(333));
+
+  // x.get() returns pointer to pointer-to object
+  int* ptr = x.get();
+  std::cout << *x << std::endl; // prints 333
+
+  // delete current pointer, stores a new one
+  x.reset(new int(124));
+  std::cout << *x << std::endl; // prints 124
+
+  // smart pointer releases responsibility for freeing back to user
+  ptr = x.release();
+  delete ptr;
+
+  return EXIT_SUCCESS;
+}
+
+```
+
+---
+## `unique_ptr` is not copyable
+
+As its name implies, `std::unique_ptr` cannot be copied
+- Copy constructor and copy-assignment operator are disabled
+- Compiler error if you try
+
+Bugged example:
+```C++
+std::unique_ptr<int> x(new int(333));
+std::unique_ptr<int> y(x); // Compiler error - cctor disabled
+
+std::unique<int> z; // z is nullptr
+z = x; // Compiler error - copy-assignment disabled
+
+```
+
+---
+## Bug #1
+
+We've seen that you can't copy unique pointers, but you can still wrap the same raw pointer with multiple unique pointers. This causes a problem
+
+Example:
+```C++
+int main() {
+    int* p = new int(333);
+    std::unique_ptr<int> x(p);
+    std::unique_ptr<int> y(p);
+
+    // What happens here at the end of main?
+}
+```
+- Now both `x` and `y` "own" pointer `p`
+- When `y` falls out of scope, it will call `delete` on `p`, which it owns
+- But then when `x` falls out of scope, it also will call `delete` on `p`, which has already been deleted!
+- Causes a double-free!
+
+---
+## Bug #2
```

Comment:
I'm actually not too sure what is meant here. Does it mean if I call delete manually on `x`, something will happen, or does it mean if `std::unique_ptr::release` or `std::unique_ptr::reset` is called. I might just be dumb here though.

Reply:
Is this about bug 2? The comment review blocks pop up at weird places sometimes

But yeah for Bug 2, if you tried to call `delete` manually on `&x`, you'd get a runtime error since it's not a pointer to heap-allocated memory

The bug in the code is basically that, but this time it's the `unique_ptr` trying to run `delete` on `&x` when the unique pointer falls out of scope

Reply:
Ahhh, I see. Then perhaps on line 153, try explicitly stating that `unique_ptr` will try to look for it in the heap because its a heap-memory based object.